File: MindFit Intelligence/v1/MindFit.Api/Services/Auth/AuthService.cs
using Microsoft.EntityFrameworkCore;
using MindFit.Api.Data;
using MindFit.Api.DTOs.Auth;
using MindFit.Api.Services.Interfaces;
using BCrypt.Net;

namespace MindFit.Api.Services.Auth;

/// <summary>
/// Servicio de autenticación con login, refresh, logout y gestión de permisos
/// </summary>
public class AuthService : IAuthService
{
    private readonly ApplicationDbContext _context;
    private readonly ITokenService _tokenService;

    public AuthService(ApplicationDbContext context, ITokenService tokenService)
    {
        _context = context;
        _tokenService = tokenService;
    }

    /// <summary>
    /// Login: Valida credenciales, genera tokens, retorna datos del usuario
    /// </summary>
    public async Task<(bool Success, string? ErrorMessage, LoginResponseDto? Response)> LoginAsync(
        LoginRequestDto request, string? ipAddress = null)
    {
        // Buscar usuario por gymId + email
        var usuario = await _context.Usuarios
            .Include(u => u.Gym)
            .Include(u => u.UsuarioGrupos)
                .ThenInclude(ug => ug.Grupo)
            .FirstOrDefaultAsync(u => u.GymId == request.GymId && u.Email == request.Email);

        if (usuario == null)
        {
            return (false, "Credenciales inválidas", null);
        }

        // Verificar si el usuario está activo
        if (!usuario.Activo)
        {
            return (false, "Usuario inactivo", null);
        }

        // Verificar si el gimnasio está activo
        if (!usuario.Gym.Activo)
        {
            return (false, "Gimnasio inactivo", null);
        }

        // Verificar contraseña con BCrypt
        if (!BCrypt.Net.BCrypt.Verify(request.Password, usuario.PasswordHash))
        {
            return (false, "Credenciales inválidas", null);
        }

        // Generar Access Token
        var accessToken = _tokenService.GenerateAccessToken(usuario);

        // Generar Refresh Token y persistirlo
        var refreshToken = _tokenService.GenerateRefreshToken(usuario.Id, ipAddress);
        await _context.SaveChangesAsync();

        // Actualizar último acceso
        usuario.UltimoAcceso = DateTime.UtcNow;
        await _context.SaveChangesAsync();

        // Obtener permisos del usuario
        var permisos = await GetUsuarioPermisosAsync(usuario.Id);

        // Construir response
        var response = new LoginResponseDto
        {
            AccessToken = accessToken,
            Usuario = new UsuarioDto
            {
                Id = usuario.Id,
                GymId = usuario.GymId,
                Email = usuario.Email,
                Nombre = usuario.Nombre,
                Apellido = usuario.Apellido,
                Telefono = usuario.Telefono,
                Grupos = usuario.UsuarioGrupos.Select(ug => ug.Grupo.Nombre).ToList()
            },
            Permisos = permisos
        };

        return (true, null, response);
    }

    /// <summary>
    /// Refresh Token: Valida refresh token, genera nuevo access token, rota refresh token
    /// </summary>
    public async Task<(bool Success, string? ErrorMessage, string? AccessToken)> RefreshTokenAsync(
        string refreshToken, string? ipAddress = null)
    {
        // Validar refresh token
        var validToken = await _tokenService.ValidateRefreshTokenAsync(refreshToken);

        if (validToken == null)
        {
            return (false, "Refresh token inválido o expirado", null);
        }

        var usuario = validToken.Usuario;

        // Verificar si el usuario está activo
        if (!usuario.Activo)
        {
            return (false, "Usuario inactivo", null);
        }

        // Generar nuevo Access Token
        var newAccessToken = _tokenService.GenerateAccessToken(usuario);

        // Generar nuevo Refresh Token (rotación)
        var newRefreshToken = _tokenService.GenerateRefreshToken(usuario.Id, ipAddress);
        await _context.SaveChangesAsync();

        // Revocar el refresh token anterior
        await _tokenService.RevokeRefreshTokenAsync(refreshToken, ipAddress, newRefreshToken.Token);

        return (true, null, newAccessToken);
    }

    /// <summary>
    /// Logout: Revoca el refresh token
    /// </summary>
    public async Task<bool> LogoutAsync(string refreshToken, string? ipAddress = null)
    {
        await _tokenService.RevokeRefreshTokenAsync(refreshToken, ipAddress);
        return true;
    }

    /// <summary>
    /// Obtiene todos los permisos de un usuario consultando sus grupos
    /// Los permisos se consultan en cada request (no se cachean en JWT)
    /// </summary>
    public async Task<List<string>> GetUsuarioPermisosAsync(int usuarioId)
    {
        var permisos = await _context.UsuarioGrupos
            .Where(ug => ug.UsuarioId == usuarioId && ug.Grupo.Activo)
            .SelectMany(ug => ug.Grupo.GrupoPermisos)
            .Where(gp => gp.Permiso.Activo)
            .Select(gp => gp.Permiso.Codigo)
            .Distinct()
            .ToListAsync();

        return permisos;
    }
}

File: MindFit Intelligence/v4/backend/src/services/auth.service.ts
import { prisma } from "../prisma.js";
import { comparePassword } from "../utils/password.util.js";
import { generateToken } from "../utils/jwt.util.js";
import type { AuthUser } from "../types/jwt.types.js";

export class AuthService {
  async login(gymId: number, nombreUsuario: string, password: string) {
    // 1. Buscar usuario en el gym especificado
    const usuario = await prisma.usuario.findUnique({
      where: {
        gymId_nombreUsuario: {
          gymId,
          nombreUsuario,
        },
      },
      include: {
        gym: true,
        persona: true,
      },
    });

    if (!usuario) {
      throw new Error("Credenciales inválidas");
    }

    // 2. Verificar contraseña
    const isValidPassword = await comparePassword(password, usuario.passwordHash);
    if (!isValidPassword) {
      throw new Error("Credenciales inválidas");
    }

    // 3. Generar JWT con payload mínimo
    const token = generateToken({
      usuarioId: usuario.usuarioId,
      gymId: usuario.gymId,
    });

    // 4. Obtener roles y permisos
    const authUser = await this.getUserWithPermissions(usuario.usuarioId, gymId);

    return {
      token,
      usuario: authUser,
    };
  }

  async getUserWithPermissions(usuarioId: number, gymId: number): Promise<AuthUser> {
    // 1. Obtener usuario con relaciones
    const usuario = await prisma.usuario.findUnique({
      where: { usuarioId },
      include: {
        gym: true,
        persona: true,
        usuarioRoles: {
          include: {
            rol: {
              include: {
                rolPermisos: {
                  include: {
                    permiso: true,
                  },
                },
              },
            },
          },
        },
        usuarioPermisos: {
          include: {
            permiso: true,
          },
        },
      },
    });

    if (!usuario || usuario.gymId !== gymId) {
      throw new Error("Usuario no encontrado o no pertenece al gym");
    }

    // 2. Extraer roles
    const roles = usuario.usuarioRoles.map((ur) => ur.rol.nombre);

    // 3. Extraer permisos de roles
    const permisosFromRoles = usuario.usuarioRoles.flatMap((ur) =>
      ur.rol.rolPermisos.map((rp) => rp.permiso.codigo)
    );

    // 4. Extraer permisos directos
    const permisosDirectos = usuario.usuarioPermisos.map((up) => up.permiso.codigo);

    // 5. Combinar y deduplicar permisos
    const permisos = [...new Set([...permisosFromRoles, ...permisosDirectos])];

    return {
      usuarioId: usuario.usuarioId,
      nombreUsuario: usuario.nombreUsuario,
      gym: {
        gymId: usuario.gym.gymId,
        nombre: usuario.gym.nombre,
      },
      persona: {
        personaId: usuario.persona.personaId,
        nombreYApellido: usuario.persona.nombreYApellido,
        email: usuario.persona.email,
      },
      roles,
      permisos,
    };
  }
}
